KAFKA-8204: fix Streams store flush order

vvcephei opened the pull request:
Streams previously flushed stores in the order of their registration,
which is arbitrary. Because stores may forward values upon flush
(as in cached state stores), we must flush stores in topological order.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-8204: fix Streams store flush order

vvcephei committed:
cleanup

vvcephei committed:
test fixes

vvcephei commented:
Hey @mjsax @cadonna,

Please take a look at this fix. I was a little stumped about the best way to enforce the topological order for flushing, I went with the current approach, in which we maintain the registered stores in sorted order during registration because I think actually registering stores isn't a hot path. Definitely open to other suggestions.

Also, as noted in the ticket (https://issues.apache.org/jira/browse/KAFKA-8204), there are other ways to fix the underlying problem. I went with this one because it doesn't require a public API change. Please also weigh in on whether you can think of any reason this fix is insufficient (or brittle) and if you think we should go for a different solution instead.

Thanks!
-John

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java:
```diff
@@ -45,7 +45,6 @@
     OffsetCheckpoint checkpoint;
 
     final Map<TopicPartition, Long> checkpointableOffsets = new HashMap<>();
-    final Map<String, StateStore> stores = new LinkedHashMap<>();
```
This was only used in ProcessorStateManager, so I moved it there and made it private.

guozhangwang replied:
+100

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -51,6 +52,8 @@
     private final Map<String, StateRestoreCallback> restoreCallbacks; // used for standby tasks, keyed by state topic name
     private final Map<String, RecordConverter> recordConverters; // used for standby tasks, keyed by state topic name
     private final Map<String, String> storeToChangelogTopic;
+    private final LinkedHashMap<String, StateStore> registeredStores = new LinkedHashMap<>();
```
This used to be `stores` in the abstract class.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -122,52 +127,64 @@ public void register(final StateStore store,
-        changelogPartitions.add(storePartition);
 
-        stores.put(storeName, store);
+        ensureStoresIsTopologicallyOrdered();
```
The alternative here is to seek to the right spot in the list and insert the new store there, but it seemed easier to read this way. Please share your thoughts.

guozhangwang replied:
Hmm I'm still not clear where did we break the topology order here: let me go through my reasoning and lmk where I got it wrong:

1. in `InternalTopologyBuilder` when we construct the `InternalTopology` the following parameter is constructed:

```
new ArrayList<>(stateStoreMap.values()),
```

So `ProcessorTopology#stateStores()` is in order.

2. in `AbstractTask#registerStateStores` we get stores from `ProcessorTopology#stateStores()` which is in order, and hence we are calling `store.init` in order, and hence call `ProcessorStateManager#register` in order as well.

3. The resulted `stores` in `ProcessorStateManager` should be in order then as well.

vvcephei replied:
I'll investigate it and let you know. I also wondered about this, but didn't look into it, since there's no guarantee that the `register` call takes place synchronously inside of `init`, so if we have a correctness requirement to flush in topo order, we should just flush in topo order, rather than depend on stores being registered in topo order, since that leaves more room for error.

vvcephei replied:
Oh, actually I see that `register` is only allowed during initialization... I'll have to go through the whole repro again to track this down. I'll let you know.

vvcephei replied:
Ah. The problem is in `org.apache.kafka.streams.processor.internals.AbstractStateManager#reinitializeStateStoresForPartitions`. We're iterating over a copy of the state stores, but when we make the copy, we don't preserve order:

```
        final Map<String, StateStore> storesCopy = new HashMap<>(stateStores);
        ...
        for (final Map.Entry<String, StateStore> entry : storesCopy.entrySet()) {
                ...
                stateStore.init(processorContext, stateStore);
                ...
        }
```

So we have alternative fix, where we fix the order of _re_-initialization, but it's not super straightforward, because not all the stores in the task get re-initialized, so even with an ordered copy of the map, it'll still wind up scrambling the stores. So, we still have the problem of explicitly maintaining topological order in the state manager.

What do you think, should we still use `ensureStoresIsTopologicallyOrdered`, or instead try to structure the initialization/re-initialization so that everything gets registered in the right order?

guozhangwang replied:
I see. So it seems this issue will only be exposed with EOS still, right (reinitializeStateStoresForPartitions is only triggered when EOS is enabled and offset out-of-range detected)?

As for the fix: since the `storesCopy = new HashMap<>(stateStores)` is just for fixing concurrent access exception, how about we just use an iterators instead, i.e.:

```
iterator = ..
for (iterator) {
  if (storeToBeReinitialized.contains(..)) {
    ...
    iterator.remove()
    stateStore.init(processorContext, stateStore);
  }
}
```

vvcephei replied:
We can do that, but I think it will still re-order the stores:
```
Stores: [A, B]
storesToBeReinitialized = [A]

iterator.next: A
storeToBeReinitialized.contains(A) == true
iterator.remove
// now stores: [B]
stateStore.init(processorContext, A)
// now stores: [B, A]
```

This is what I meant by saying that we still have the problem of maintaining topological order during registration, even if we correct the ordering of `storesCopy`. We could add some mechanism to keep it in the same position in the linked map, but part of me feels like the current code in the PR is better, since it just enforces the desired invariant on the stores after every registration, regardless of what order things happen in.

guozhangwang replied:
I see your point now, this is exactly the messy code that we were trying to fix.

I've looked at the source code again, and I think we can actually not remove the state at all since the same object will be add to the state stores via `store.init` immediately within the same function call.

So I think we can actually do:

```
if (storeToBeReinitialized.contains(A)) {
A.close;
delete state dir;
A.init();
}
```

In that loop.

vvcephei replied:
Yes, that would work.

vvcephei replied:
Ok, please see my latest comment below. The fly in the ointment was that we don't allow to register a store twice, and the only way we know whether a store is registered is its presence in this map.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:
```diff
@@ -34,77 +35,19 @@
     private final Map<String, String> storeToChangelogTopic;
     private final Set<String> repartitionTopics;
 
-    public static ProcessorTopology with(final List<ProcessorNode> processorNodes,
```
These factories were only used by tests, so it's not appropriate to have them in the main module. Their presence here made it more difficult to analyse the production code paths that construct a ProcessorTopology.

So, I relocated these factory methods to the test module. The ones that were used in exactly one test got moved to that test class, and the others got moved to a `ProcessorTopologyFactories` test utility.

guozhangwang replied:
+100

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:
```diff
@@ -225,4 +168,8 @@ public String toString(final String indent) {
     }
+
+    public List<String> topologicalOrderOfStores() {
+        return stateStores.stream().map(StateStore::name).collect(Collectors.toList());
```
`stateStores` is already topologically sorted, so I'm just transforming it to a list of store names.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/AbstractTaskTest.java:
```diff
@@ -96,6 +97,8 @@ public void shouldThrowWakeupExceptionOnInitializeOffsetsWhenWakeupException() {
     public void shouldThrowLockExceptionIfFailedToLockStateDirectoryWhenTopologyHasStores() throws IOException {
         final Consumer consumer = EasyMock.createNiceMock(Consumer.class);
         final StateStore store = EasyMock.createNiceMock(StateStore.class);
+        expect(store.name()).andReturn("dummy-store-name").anyTimes();
```
This is kind of funny. We always called `store.name()`, but with the current changes, we call it twice. Apparently, EasyMock returns `null` on the first invocation and throws an exception on the second, which doesn't seem quite right.

Anyway, since we need the store to have a name, I made one.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/AbstractTaskTest.java:
```diff
@@ -232,9 +235,13 @@ private AbstractTask createTask(final Consumer consumer,
+                                new StoreChangelogReader(consumer,
+                                                         Duration.ZERO,
+                                                         new MockStateRestoreListener(),
+                                                         new LogContext("stream-task-test ")),
```
I also shorted some long lines.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -362,6 +368,9 @@ public void testFlushAndClose() throws IOException {
         assertTrue(checkpointFile.exists());
 
+        // make sure that flush is called in the proper order
+        assertThat(persistentStore.getLastFlushCount(), Matchers.lessThan(nonPersistentStore.getLastFlushCount()));
```
Here's the regression test for my code change. Previously, we would flush the stores in registration order, but now we do it in topological order (defined on L346).

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -517,7 +533,8 @@ public void shouldThrowIllegalArgumentExceptionOnRegisterWhenStoreHasAlreadyBeen
             stateDirectory,
-            Collections.emptyMap(),
+            emptyMap(),
+            singletonList(mockKeyValueStore.name()),
```
note that stores need not be represented in the `name->changelog` map, but they must be listed in the topological order.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -658,6 +668,17 @@ public void flush() {
+            changelogReader,
+            false,
+            logContext);
+
```
I had to move a couple of these so that I could list the store names in the topological order (I.e., I have to generate the names before I can list them).

vvcephei commented:
Test failed with environmental error:
```
ERROR: Failed to write output for test null.Gradle Test Executor 5
java.lang.NullPointerException: Cannot invoke method write() on null object
	at org.codehaus.groovy.runtime.NullObject.invokeMethod(NullObject.java:91)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:47)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.NullCallSite.call(NullCallSite.java:34)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at java_io_FileOutputStream$write.call(Unknown Source)
	at build_35h9280v69j3c1fqne09q9xlz$_run_closure5$_closure74$_closure107.doCall(/home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/build.gradle:244)
	at sun.reflect.GeneratedMethodAccessor269.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:411)
	at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:40)
	at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:25)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:324)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:234)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy62.onOutput(Unknown Source)
	at org.gradle.api.internal.tasks.testing.results.TestListenerAdapter.output(TestListenerAdapter.java:56)
	at sun.reflect.GeneratedMethodAccessor266.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:324)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:234)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy63.output(Unknown Source)
	at org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor.output(StateTrackingTestResultProcessor.java:112)
	at org.gradle.api.internal.tasks.testing.results.AttachParentTestResultProcessor.output(AttachParentTestResultProcessor.java:48)
	at sun.reflect.GeneratedMethodAccessor268.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.FailureHandlingDispatch.dispatch(FailureHandlingDispatch.java:29)
	at org.gradle.internal.dispatch.AsyncDispatch.dispatchMessages(AsyncDispatch.java:87)
	at org.gradle.internal.dispatch.AsyncDispatch.access$000(AsyncDispatch.java:36)
	at org.gradle.internal.dispatch.AsyncDispatch$1.run(AsyncDispatch.java:71)
	at org.gradle.internal.concurrent.InterruptibleRunnable.run(InterruptibleRunnable.java:42)
	at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

vvcephei commented:
Retest this, please.

guozhangwang reviewed (commented):
Made a pass on the PR. I looked over the source code again but still wondering where did we break the order.

vvcephei committed:
cr comment to preserve registration order

vvcephei commented:
@guozhangwang , I've taken your feedback on preserving the stores order, rather than enforcing it after registration.

Please take a look at that latest commit, and let me know what you think. If you're happy with it, I'll make another pass and clean up the whole PR.

It wound up being a little bit complicated. I assume we want to preserve the restriction that stores can't be registered twice, so I had to wrap the values in the map with `Optional` (there are other ways to do it, but it all amounts to the same thing). The implementation would be quite vulnerable to future ordering bugs, so I added a restriction that once added, a mapping cannot be removed. This should safeguard against anything like this happening again. Unfortunately, because of the shared logic among ProcessorStateManager, GlobalStateManager, and AbstractStateManager, all three classes have to change.

WDYT?

guozhangwang commented:
@vvcephei I've taken a look at the latest commit and I like it better than the original approach: just make the ordering un-modifiable while allowing the object to be replaced -- it's unfortunate that we use the map for registration check but it sort of also necessary to make sure the same store is not registered twice. Note the only case it may happen is with EOS + offset changed or sth. else that causes re-init. And the fact that we are a bit messy around this code base (assigned tasks -> processor state managers -> impl classes) is also making things worse. We have a ticket open for cleaning this tech debt as soon as possible.

One nit thing is that upon task initialization (e.g. `StreamTask#initializeTopology`), we should check inside processor state manager that all states are not `empty`: of course it should be the case if we are bug-free, but since the code around these areas are quite messy I'd like to also add this safeguard as well. Otherwise I'm +1 on the approach.

vvcephei committed:
sanity check

vvcephei committed:
cleanup

vvcephei committed:
cleanup

vvcephei commented:
Ok, @guozhangwang, I think this PR is ready for another full look now.

On cleanup, I noticed that, since the "reinitialize" method can't change the order of the stores anymore, there's no need to iterate over all stores, we only need to iterate over the stores we're going to reinitialize. I'd appreciate a good look at the logic there to make sure I didn't mess anything up.

Thanks!

vvcephei committed:
suppress deprecation warnings in FixedOrderMapTest

vvcephei committed:
fix style

vvcephei committed:
fix tests

vvcephei commented:
Hi @guozhangwang , the tests have passed and all the comments are addressed. Is this ready to merge?

guozhangwang commented:
Sorry for being late! Will review again.

On Wed, Apr 17, 2019 at 3:26 PM John Roesler <notifications@github.com>
wrote:

> Hi @guozhangwang <https://github.com/guozhangwang> , the tests have
> passed and all the comments are addressed. Is this ready to merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/6555#issuecomment-484284189>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACR3Q7NGBJYLS3ZMTGXJ6X3PQ6PZFANCNFSM4HEU775A>
> .
>


-- 
-- Guozhang

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java:
```diff
@@ -82,36 +80,44 @@ public void reinitializeStateStoresForPartitions(final Logger log,
-                stateStore.init(processorContext, stateStore);
+        for (final String storeName : storesToBeReinitialized) {
+            if (!stateStores.containsKey(storeName)) {
+                // the store has never been registered; carry on...
```
Is this a normal situation? At the moment we will only re-initialize if it was already registered, I'd suggest we treat it as an IllegalStateException

vvcephei replied:
I saw it happen during tests. I'm not sure if it's a testing artifact or not. Note that we would never have encountered the situation before, since we were only iterating over registered stores before.

guozhangwang replied:
Would be interest to know what you could find out: my gut feeling it is was not testing artifact :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/FixedOrderMap.java:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Maybe we can put this class in `o.a.k.common.utils` instead, so it can be used elsewhere?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java:
```diff
@@ -45,8 +45,7 @@
     final Map<TopicPartition, Long> checkpointableOffsets = new HashMap<>();
-    final Map<String, StateStore> stores = new LinkedHashMap<>();
-    final Map<String, StateStore> globalStores = new LinkedHashMap<>();
+    final FixedOrderMap<String, Optional<StateStore>> globalStores = new FixedOrderMap<>();
```
Could we move `globalStores` to GlobalStateManagerImpl as well?

vvcephei replied:
No, because the ProcessorStateManager offers an interface to look up global stores, as well (for some reason, maybe joins?)

guozhangwang replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -306,12 +307,15 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
-                store.flush();
-            } catch (final Exception e) {
-                throw new ProcessorStateException(String.format("Failed to flush global state store %s", store.name()), e);
+        for (final Optional<StateStore> maybeStore : globalStores.values()) {
```
After the task has been transited to `running`, it should mean that all its state stores have been successfully initialized and restored, so this should not happen. If it ever happens we should treat it as IllegalState indicating Streams bug. 


Ditto elsewhere.

guozhangwang replied:
Actually thinking about it: today we can reinitialize in three cases:

1) for active / standby / global task, when offset out of range during `restoring`.
2) for standby task, when offset out of range during `running`.
3) for active / standby / global task, when checkpoint not found and EOS enabled when starting `restoring`.

I.e. it seems to me that for any `task#flush / checkpoint` call should have all `present` stores, while for `task#close` some stores may not be `present` if an exception in the store#initialize caused the whole task to be closed. So I suggest we throw Illegal for the former functions, while logging an INFO for the latter.

vvcephei replied:
I'll implement your suggestions when I get a chance.

I can see the high-level value of tightening our assertions about what state these things can be in, to catch more bugs. But on the other hand, does it make sense for the StateManager to be that opinionated about the lifecycle of the Task that contains it?

vvcephei replied:
One note, the code I've submitted is functionally equivalent to the pre-existing code, since under the situation where a `maybeStore` is `empty`, in the existing codebase, the mapping would simply be absent from the `stores` map and therefore skipped in this iteration.

guozhangwang replied:
Yeah that's true. I was trying to tighten the screw a bit in case we have some lurking bugs, it would be easier to detect.

That being said, I think I also agree having opinionated StateManager depending on other classes is not a good design pattern. So how about taking one step back from aggressiveness and just log it when we are skipping on these functions?

vvcephei replied:
Cool. I didn't get a chance to look at it yet, but I'll see how I can best accomplish the goal.

vvcephei replied:
Ok, I looked around, and I agree it should be fine to put in those else conditions. Let's see if the tests pass! ;)

guozhangwang reviewed (commented):
@vvcephei I made another pass and left some minor comments about `isPresent` false case handling. Otherwise it LGTM.

vvcephei committed:
move FixedOrderMap

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java:
```diff
@@ -82,36 +80,44 @@ public void reinitializeStateStoresForPartitions(final Logger log,
+                    "Re-initializing store that has not been initialied. This is a bug in Kafka Streams."
+                );
+            }
+            final StateStore stateStore = maybeStore.get();
```
super nit: replace lines 89-94 with 
```java
final StateStore stateStore = maybeStore.orElseThrow(() -> new IllegalStateException(
                "Re-initializing store that has not been initialied. This is a bug in Kafka Streams."
            ));
```

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java:
```diff
@@ -45,8 +46,7 @@
     final Map<TopicPartition, Long> checkpointableOffsets = new HashMap<>();
-    final Map<String, StateStore> stores = new LinkedHashMap<>();
-    final Map<String, StateStore> globalStores = new LinkedHashMap<>();
+    final FixedOrderMap<String, Optional<StateStore>> globalStores = new FixedOrderMap<>();
```
Minor comment after looking over the whole PR, do we need a `FixedOrderMap<String, Optional<StateStore>>` vs just `FixedOrderMap<String, StateStore>`?

IIRC we can't pass in a `null` state store and from what I can see it's the `GlobalStateManager` and `ProcessorStateManager` that put an `Optional.empty()` in the map vs. clearing the map in the respective `close` methods.

Not a big deal, IMHO it added a little complexity to the code and I'm not sure what we gain with it.

vvcephei replied:
I'm sorry, but I'm not quite getting the feedback. The GlobalStateManager passes this collection into `AbstractStateManager#reinitializeStateStoresForPartitions`, which may de-register some of the stores, just like what happens with the ProcessorStateManager. Or did you mean something else?

bbejeck reviewed (commented):
Thanks for this PR @vvcephei I've made a pass and looks good overall.

vvcephei committed:
cr feedback

vvcephei committed:
cr feedback

vvcephei commented:
Ok, @bbejeck and @guozhangwang , please take another look at this PR. I think I've addressed all the feedback so far (mod the one I asked a question about). Thanks for the reviews!

vvcephei committed:
Close is no longer idempotent

vvcephei commented:
@guozhangwang , one test broke as a result of that screw tightening... As a consequence of the new check, closing the state manager is no longer idempotent (see https://github.com/apache/kafka/pull/6555/commits/d4a68c6929c40cf6d78e69d53bc6fe4a57527383) . I don't think this is a problem; WDYT?

guozhangwang commented:
Thanks @vvcephei . I agree with you that for this unit test, the above change makes sense: philosophically a store (or state manager that owns a bunch of stores) should not be flushed after it was closed.

As for the production safety, I'd suggest we move line 280 for de-register the store after store.close() to be within the try-block as well: i.e. if a store.close() throws exception we will not de-register it.

Other than this, I'm +1 on the PR.

vvcephei committed:
move deregistration into try block

vvcephei commented:
Done! Thanks for the reviews, @guozhangwang and @bbejeck. Are there any final comments, or is this just waiting for the tests to pass now?

guozhangwang commented:
Merged to trunk and cherry-picked with minor conflicts resolved to 2.2.

Merging to 2.1 resulted in too many conflicts..

vvcephei commented:
Thanks, @guozhangwang ! I'll send a PR later with the cherry-pick to 2.1
